### PR TITLE
[MU4] Fix #12925: Dock panel does not respect maximum size anymore after moving it from leftmost column to other location

### LINKED ISSUE #12925: [MU4 Issue] Dock panel does not respect maximum size anymore after moving it from leftmost column to other location
**To Reproduce**
1. Open a score on the "Score" page.
2. Have some panels docked in the leftmost column.
3. Move one or more panels to a different dock location in a different column.
See error: the moved panel does not respect its maximum size anymore; you can resize it very wide.
4. Open the Mixer 
See error: the moved panel becomes automatically very wide (it seems that the space is distributed equally between the notation view and the moved panel)
5. Switch to another page and switch back to the "Score" page.
Note that everything is fine again, until you move a panel again to the left column and then again move it to another location.

**Expected behavior**
Maximum size should be respected.

**Screenshots**

https://user-images.githubusercontent.com/48658420/185448445-9f8ca1b4-38b9-45e7-9d55-6eac4e429464.mov

**Platform information**
 - OS: reported on macOS

### PR DESCRIPTION
**Pull request by RomanPudashkin**
Resolves: https://github.com/musescore/MuseScore/issues/12925
Resolves: https://github.com/musescore/MuseScore/issues/12806
Resolves: https://github.com/musescore/MuseScore/issues/12789
Resolves: https://github.com/musescore/MuseScore/issues/12976

Reworked [my solution](https://github.com/musescore/MuseScore/pull/12431/commits/d89f2cb6146b0d51605aecca7515c7a097d3b825) for the bug https://github.com/musescore/MuseScore/issues/12421, because it doesn't cover some cases

### CODE REVIEW

File: src/appshell/view/dockwindow/internal/dockbase.cpp [hunk 1]
```diff
@@ -640,6 +642,12 @@ void DockBase::updateFloatingStatus()
     doSetFloating(floating);
 }
 
+void DockBase::onIsInMainWindowChanged()
+{
+    applySizeConstraints();
+    updateFloatingStatus();
```
Comment (cbjeukendrup) [inline src/appshell/view/dockwindow/internal/dockbase.cpp:648]: Is this method executed only when `isInMainWindow` has changed to `true`, or also when it has changed to `false`? If the latter, isn't this duplicating the code in line +621 to +624, or is that code invoked in different situations?

Or, to rephrase my question: when is `KDDockWidgets::Frame::isInMainWindowChanged` emitted and when `KDDockWidgets::DockWidgetQuick::parentChanged`?

**Review by cbjeukendrup — APPROVED** [review]
Looks good! I have one question though.

File: src/appshell/view/dockwindow/internal/dockbase.cpp (hunk 1 above)
Comment (RomanPudashkin) [reply src/appshell/view/dockwindow/internal/dockbase.cpp:648]: >Is this method executed only when isInMainWindow has changed to true, or also when it has changed to false? 

It is called in both cases

>If the latter, isn't this duplicating the code in line +621 to +624, or is that code invoked in different situations?

The code in lines 621-624 is needed for one particular case: when the dock becomes undocked and loses its origin frame. KDDockWidgets::Frame::isInMainWindowChanged won't be called in this case (we just haven't yet connected to this signal from a new frame). This code ensures that the size constraints will be applied to the floating dock

Comment (RomanPudashkin) [reply src/appshell/view/dockwindow/internal/dockbase.cpp:648]: PS: docks do not always change their parent when they are undocked. Several docks can use the same frame at the same time. And that frame really belongs to only one of them (the first one who created it)

Comment (cbjeukendrup) [reply src/appshell/view/dockwindow/internal/dockbase.cpp:648]: Ah, thanks for the clarification!

**Comment by DmitryArefiev** [comment]
Tested #12925 #12806 #12806 #12976 on Win10, Mac12, LinuxMint 20.2 - **FIXED**

**Pull request merged**